### ISSUE TITLE
FBX importer: Banshee cannot find all the bones when importing some models

### ISSUE THREAD

**Issue by christianclavet** [BearishSun/BansheeEngine#166]
Hi,

I'm trying to import "HANDS" with animations with Banshee and still having issues. I'm having better results since the last changes, but the animation playback is not working because Banshee thinks there is something missing in the skeleton hierarchy.

The file that I've got came from Blender and I had a very difficult time importing it. Tried some things and was never sure of what happened (more experienced with 3DS Max)

Finally I was able to export something useful that MAX could read without crashing.
The file is FPS_Hands_V2.fbx (imported in 3DS Max), after some tweaking were able to see something useful like this and the animation work (imported a single animation), the FBX contain a lots! (I had to rotate on the X axis)
![3dsmax](https://user-images.githubusercontent.com/13395943/35831676-933c5990-0a98-11e8-8008-8d003b84ff33.JPG)

Then exported this as a new FBX (**HandTest.FBX**)
And got the model ok in Banshee, but got this error:

> Not all bones were found in the node hiearchy. Skeleton invalid.

![banshee](https://user-images.githubusercontent.com/13395943/35831748-ce2c6586-0a98-11e8-93f4-26f2a6ee2df6.JPG)
And there is no animation, but I don't see stretching.

If you need checking the sources:
The .blend original file and the 2 FBX (first exported by Blender, the other by 3DS Max)
[FPS Hands.zip](https://github.com/BearishSun/BansheeEngine/files/1696983/FPS.Hands.zip)

I've rechecked my other assets (mostly the ship and the "Merc" character and they import and display correctly. There is something with this one that cause the problem.

**EDIT:** 
Found the problem for the "bones not found" .  An option is the FBX  blender exporter should not be used:
"Use leaf bone" - To add a bone at the end of the chain for better editing. But seem that Banshee does'nt like it.
I have other test rigs here that produce FBX that exibit this problem (3DS Max "Cat rigs" or Biped)

I was able to export back from Max and this time, the animation played back. But loading the exported .FBX (binary) that was done by Blender gave a problem when trying to load the animation data from it. Banshee report this:

> "illegal characters in path"

I then decided to export the animation again from Blender as FBX Ascii. The base orientation seem wrong, but I can see all the animations and there no stretching.
![capture](https://user-images.githubusercontent.com/13395943/35837300-316dfdf2-0ab3-11e8-8982-dcb919eb672b.JPG)

[FPS Hands_update.zip](https://github.com/BearishSun/BansheeEngine/files/1697361/FPS.Hands_update.zip)

Here are the updated files if you need the check the source FBX. And reproduce the issue.
Note: There was lot of options to set for a proper export from Blender and the scene had multiple layers that I needed to deselect. But If 3DS Max could load it, then Banshee was able also being the only exception is the FBX Binary.  The file load correctly in Max but Banshee will fail to import it (when animation is checked)..

Note: Is there a way to parent to a bone directly from the editor without using a script? Seen the bone component, but just expose a bone so we can use a script to parent an object to it. I was expecting something like Unity is that you expand the node to see the bone and drag the object there. But fail to find this in the editor...

**Comment by BearishSun** [comment]
Thanks. I'll check what the issues are with those FBX files, they certainly shouldn't be happening.

You don't need to use scripting in order to use the Bone component. It works similar to Unity, except you need to manually create a scene object that's a child of the object containing the Animation component. You then attach a Bone component to it, and select which bone it influences. At that point you can add objects as the children of that object, and it should animate together with the bone (provided this works, I haven't tested it after recent animation system changes).

**Comment by christianclavet** [comment]
Hi! Cool! Just tested the bone component and it doesn't work for now. I'm able to select the desired bone with the GUI but it reset all the time. My object is not moving.

**Comment by BearishSun** [comment]
The following issues should now be resolved:
 - "illegal characters in path" when animation names have invalid characters
 - "Not all bones were found in the node hierarchy. Skeleton invalid." when multiple meshes share a skeleton
 - Bone attachments failing to animate properly

I believe I've tested with all the models you've sent me, and animations appear to import and play correctly. Closing the issue, but feel free to reopen if I missed something.

**Comment by christianclavet** [comment]
Tested, and I all seem to work.
For the attach to bone, as you said I used a generic sceneobject parented to the object, selected the bone to use then parented the object I was to parent it with.
![capture](https://user-images.githubusercontent.com/13395943/36331570-65bf75e0-133c-11e8-95da-de898867afca.JPG)
It was really hard to figure the proper positioning and rotation for placing the weapon.
1. Entering values in the numeric input didnt worked, values entered in X for example were affecting the other numeric input axis. (rotation)
2. Since the generic scene object is not parented and oriented when we do it from the GUI. I tried to place it manually and rotate with the Gizmo instead and here is the result.

It would be nice that once you select the bone in the GUI, it would position and orientate the object so it easier to position the desired object afterward...
